rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,410 @@
+---
+authors: Stephen Levine (stephen.levine@goteleport.com)
+state: draft
+---
+
+# RFD 0169 - Automatic Updates for Linux Agents
+
+## Required Approvers
+
+* Engineering: @russjones && @bernardjkim
+* Product: @klizhentas || @xinding33 
+* Security: Vendor TBD
+
+## What
+
+This RFD proposes a new mechanism for Teleport agents installed on Linux servers to automatically update to a version set by an operator via tctl.
+
+The following anti-goals are out-of-scope for this proposal, but will be addressed in future RFDs:
+- Analogous adjustments for Teleport agents installed on Kubernetes
+- Phased rollouts of new agent versions for agents connected to an existing cluster
+- Signing of agent artifacts via TUF
+- Teleport Cloud APIs for updating agents
+
+This RFD proposes a specific implementation of several sections in https://github.com/gravitational/teleport/pull/39217.
+
+Additionally, this RFD parallels the auto-update functionality for client tools proposed in https://github.com/gravitational/teleport/pull/39805.
+
+## Why
+
+The existing mechanism for automatic agent updates does not provide a hands-off experience for all Teleport users.
+
+1. The use of system package management leads to interactions with `apt upgrade`, `yum upgrade`, etc. that can result in unintentional upgrades or confusing command output.
+2. The use of system package management requires complex logic for each target distribution.
+3. The installation mechanism requires 4-5 commands, includes manually installing multiple packages, and varies depending on your version and edition of Teleport.
+4. The use of bash to implement the updater makes changes difficult and prone to error.
+5. The existing auto-updater has limited automated testing.
+6. The use of GPG keys in system package managers has key management implications that we would prefer to solve with TUF in the future.
+7. The desired agent version cannot be set via Teleport's operator-targeted CLI (tctl).
+8. The rollout plan for the new agent version is not fully-configurable using tctl.
+9. Agent installation logic is spread between the auto-updater script, install script, auto-discovery script, and documentation.
+10. Teleport contains logic that is specific to Teleport Cloud upgrade workflows.
+11. The existing auto-updater is not self-updating.
+12. It is difficult and undocumented to automate agent upgrades with custom automation (e.g., with JamF). 
+
+We must provide a seamless, hands-off experience for auto-updates that is easy to maintain.
+
+## Details
+
+We will ship a new auto-updater package written in Go that does not interface with the system package manager.
+It will be distributed as a separate package from Teleport, and manage the installation of the correct Teleport agent version manually. 
```

Comment:
We've hit a couple of additional issues in the past week that have been indirectly caused by us adding more and more to the Teleport release process, rather than treating tools as separate "products" as discussed here:
* Our artifact size is perpetually growing. I did some estimations for an unrelated RFD, and the total cost to the company for each additional MB of artifact _**for OS packages alone**_ is around $1000 per year. This doesn't sound like a lot, but when we're unable to produce new binaries less than 15-30MB (due to language + library choices), this adds up quickly. _**Again, this is for artifacts that end up in OS package repos only. The total cost to the company is significantly higher**_. I would guess that the total cost to the company for each new MB of artifact that goes through our normal release process is at least 3x - 5x this cost, which makes the cost of versioning this with Teleport _**at least**_ $150k+/year.

  [We're all shocked when we see that upcoming releases contain well over a gigabyte of binaries](https://gravitational.slack.com/archives/C0DF0TPMY/p1717546388403979). Things like this are what's driving this issue.
* Our build and release pipelines are already really complex, and each new additional binary that we add compounds this. As discussed [here](https://gravitational.slack.com/archives/C0351AUPJ8K/p1717684315363659), it's already a large amount of work just to get a single new binary released with Teleport. This makes it harder to iterate quickly, and to complete work that our customers need. An argument has been made in the past that we only need to do this infrequently, but if you look at the past year, we've actually increased the number of release artifacts we have significantly.

Reply:
The issue [we just hit](https://gravitational.slack.com/archives/C04P78M46F2/p1718125441924689) with one of our customers is another example of why we should version/release the updater separately. The more we couple together, the larger the impact when things go wrong.